### PR TITLE
feat: Added custom option and config

### PR DESCRIPTION
**Pull request by ElrohirGT**
This PR resolves (https://github.com/cachix/devenv/issues/1212) by adding a configuration option that enables the user to start the PostgreSQL server with a custom pg_hba.conf file, this file is primarily used to enable or disable which IPs get to connect to the server both locally (from UNIX sockets) or from the network (using TCP/IP).

I had a "Fine I'll do it myself moment" and honestly after I saw the code it wasn't really that hard.

### CODE REVIEW

File: src/modules/services/postgres.nix [hunk 1]
```diff
@@ -276,6 +281,19 @@ in
         CREATE ROLE bar;
       '';
     };
+
+    customPgHbaFile = lib.mkOption {
```
Comment (domenkozar) [inline src/modules/services/postgres.nix:285]: I'd name it `hbaConf` and make it accept a string, for consistency with other options.

**Comment by domenkozar** [comment]
> I had a "Fine I'll do it myself moment" and honestly after I saw the code it wasn't really that hard.

:muscle: :muscle: :muscle:

File: src/modules/services/postgres.nix (hunk 1 above)
Comment (ElrohirGT) [reply src/modules/services/postgres.nix]: Your right! Done.

File: src/modules/services/postgres.nix [hunk 2]
```diff
@@ -85,6 +85,8 @@ let
 
   configFile = pkgs.writeText "postgresql.conf" (lib.concatStringsSep "\n"
     (lib.mapAttrsToList (n: v: "${n} = ${toStr v}") cfg.settings));
+  setupPgHbaFileScript = if cfg.hbaConf != null then
+    ''cp ${cfg.hbaConf} "$PGDATA/pg_hba.conf"'' else "";
```
Comment (domenkozar) [inline src/modules/services/postgres.nix:89]: You'll need to use `pkgs.writeText here to get it into Nix store.

File: src/modules/services/postgres.nix [hunk 3]
```diff
@@ -276,6 +281,19 @@ in
...
       '';
     };
+
+    hbaConf = lib.mkOption {
+      type = types.nullOr types.str;
+      default = null;
+      description = ''
+        The path to a custom pg_hba.conf file to copy into the postgres installation. This
+        allows for custom connection rules that you want to establish on the server. If
+        you're using flakes remember to add the file to version control!
```
Comment (domenkozar) [inline src/modules/services/postgres.nix:291]: The flake remark no longer applies.

File: src/modules/services/postgres.nix [hunk 4]
```diff
@@ -276,6 +281,19 @@ in
...
+      '';
+      example = lib.literalExpression ''
+        "${./my-custom/directory/to/pg_hba.conf}"
```
Comment (domenkozar) [inline src/modules/services/postgres.nix:294]: Needs to use `builtins.readFile`

**Review by domenkozar — CHANGES_REQUESTED** [review]

File: src/modules/services/postgres.nix (hunk 4 above)
Comment (ElrohirGT) [reply src/modules/services/postgres.nix:294]: Isn't just putting the file path enough to copy it into the Nix store? From what I understand it copies the file into a path in the nix store and then just references it.

Source: https://stackoverflow.com/questions/43850371/when-does-a-nix-path-type-make-it-into-the-nix-store-when-converted-to-a-string

Comment (domenkozar) [reply src/modules/services/postgres.nix:294]: Yeah but it expect it a string now, so you don't want the contents of pg_hba.conf to be a nix store path, but the contents of it.

Comment (ElrohirGT) [reply src/modules/services/postgres.nix:294]: Gotcha, ok! I'll do the changes in a minute

**Comment by domenkozar** [comment]
Need to run `pre-commit run -a`.

**Comment by domenkozar** [comment]
That's strange. You cab just run nixpkgs-fmt on the file you change as a workaround

**Comment by ElrohirGT** [comment]
Gotcha! [There's two nixfmt](https://search.nixos.org/packages?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=nixfmt) on the store, I tried both. Decided on the one with the less amount of changes to the rest of the file.

**Comment by domenkozar** [comment]
We're using nixpkgs-fmt

**Comment by ElrohirGT** [comment]
> We're using nixpkgs-fmt

haha gotcha, I'm dumb, I thought you were refering to nixfmt. Fixed it! Now it should be on the correct format.

**Comment by ElrohirGT** [comment]
Hmmm... weird. I don't know why the tests fail on Mac ARM-64 but I can confirm that using this code I can run it successfully on my project.

**Comment by domenkozar** [comment]
I believe that's an error on our side.

**Pull request merged**